imap_processing/ultra/l1a/ultra_l1a.py

```diff
@@ -256,33 +287,51 @@ def create_dataset(decom_ultra_dict: dict):
     return dataset
 
 
-def ultra_l1a(packet_file: Path, xtce: Path):
+def ultra_l1a(packet_file: Path, apid: Optional[int] = None):
     """
     Process ULTRA L0 data into L1A CDF files at output_filepath.
 
     Parameters
     ----------
     packet_file : dict
         Dictionary containing paid and path to the CCSDS data packet file.
-    xtce : Path
-        Path to the XTCE packet definition file.
+    apid : int, optional
+        Optional apid
+
+    Returns
+    -------
+    output_datasets : list of xarray.Dataset
+        List of xarray.Dataset
     """
-    if ULTRA_EVENTS.apid[0] in packet_file.keys():
-        # For events data we need aux data to calculate event times
-        apid = ULTRA_EVENTS.apid[0]
-        decom_ultra_events = decom_ultra_apids(packet_file[apid], xtce, apid)
-        decom_ultra_aux = decom_ultra_apids(
-            packet_file[ULTRA_AUX.apid[0]], xtce, ULTRA_AUX.apid[0]
-        )
-        decom_ultra_dict = {
-            ULTRA_EVENTS.apid[0]: decom_ultra_events,
-            ULTRA_AUX.apid[0]: decom_ultra_aux,
-        }
+    xtce = Path(
+        f"{imap_module_directory}/ultra/packet_definitions/" f"ULTRA_SCI_COMBINED.xml"
+    )
+
+    packets = decom.decom_packets(packet_file, xtce)
+    grouped_data = group_by_apid(packets)
+
+    output_datasets = []
+
+    if apid is not None:
+        apids = [apid]
     else:
-        apid = next(iter(packet_file.keys()))
-        decom_ultra_dict = {apid: decom_ultra_apids(packet_file[apid], xtce, apid)}
+        apids = grouped_data.keys()
+
+    for apid in apids:
+        data = {apid: grouped_data[apid]}
+
+        if apid == ULTRA_EVENTS.apid[0]:
+            data_aux = {ULTRA_AUX.apid[0]: grouped_data[ULTRA_AUX.apid[0]]}
+
+            decom_ultra_dict = {
+                apid: decom_ultra_apids(data, apid),
```

Comment:
can you rename `decom_ultra_apids` to something else? You already decom on line 310 and now you are processing it.
```suggestion
                apid: process_ultra_apids(data, apid),
```

post SIT-3, I think it will be helpful to refactor this function and create_dataset into it's separate components.
```
if aux apid:
   process aux apid
         process aux apid and save it to its dataset
elif event apid:
   process event apid
         process event apid and save it to its dataset
so on
```
This will help to reduce if and else statement in many places and it will make update to certain packet easier.